staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -1043,10 +1053,20 @@ func (c *cacheWatcher) process(initEvents []*watchCacheEvent, resourceVersion ui
 
 	defer close(c.result)
 	defer c.Stop()
-	for event := range c.input {
-		// only send events newer than resourceVersion
-		if event.ResourceVersion > resourceVersion {
-			c.sendWatchCacheEvent(event)
+	for {
+		select {
+		case event, ok := <-c.input:
+			if !ok {
+				return
+			}
+			// only send events newer than resourceVersion
+			if event.ResourceVersion > resourceVersion {
+				c.sendWatchCacheEvent(event)
+			}
+		case <-ctx.Done():
```

Comment:
Hmm - I started wondering about this one.
Given that we want for "ctx.Done", and we compute timeout purely based on "ctx.Deadline()", do we even need timer here?
Can't we purely rely on "ctx.Done"?

Reply:
ctx.Done is triggered externally(for cancellation), the timer is triggered internally(for timeout).

Reply:
but when timeout (attached to context) fires, the ctx.Done is also closed:

So what's the difference?

Reply:
I mean, do we really need to differentate ctx.Done() from timeout (that was computed based on context.Deadline() anyway)?

Reply:
leave the ctx.Done which will be triggered by the timer in handler/watch.go

Reply:
looks good - thanks